src/safemode_ui.cpp

```diff
@@ -0,0 +1,740 @@
+            if( ( tab == GLOBAL_TAB && g->u.name != "" ) || tab == CHARACTER_TAB ) {
+                changes_made = true;
+                //copy over
+                auto &temp_rules_from = ( tab == GLOBAL_TAB ) ? global_rules : character_rules;
+                auto &temp_rules_to = ( tab == GLOBAL_TAB ) ? character_rules : global_rules;
+
+                temp_rules_to.push_back( rules_class(
+                                             temp_rules_from[line].rule,
+                                             temp_rules_from[line].active,
+                                             temp_rules_from[line].exclude,
+                                             temp_rules_from[line].attitude,
+                                             temp_rules_from[line].proximity ) );
+
+                //remove old
+                temp_rules_from.erase( temp_rules_from.begin() + line );
+                line = temp_rules_from.size() - 1;
+                tab = ( tab == GLOBAL_TAB ) ? CHARACTER_TAB : GLOBAL_TAB;
+            }
+        } else if( action == "CONFIRM" && current_tab_non_empty ) {
+            changes_made = true;
+            if( column == COLUMN_RULE ) {
+                fold_and_print( w_help, 1, 1, 999, c_white,
+                                _(
+                                    "* is used as a Wildcard. A few Examples:\n"
+                                    "human          matches every NPC"
+                                    "zombie         matches the monster name exactly\n"
+                                    "acidic zo*     matches monsters beginning with 'acidic zo'\n"
+                                    "*mbie          matches monsters ending with 'mbie'\n"
+                                    "*cid*zo*ie     multiple * are allowed\n"
+                                    "AcI*zO*iE      case insensitive search\n"
+                                    "" )
+                              );
+
+                draw_border( w_help );
+                wrefresh( w_help );
+                current_tab[line].rule = wildcard_trim_rule( string_input_popup( _( "Safemode Rule:" ),
+                                         30, current_tab[line].rule ) );
+            } else if( column == COLUMN_INCLUDE_EXCLUDE ) {
+                current_tab[line].exclude = !current_tab[line].exclude;
+            } else if( column == COLUMN_ATTITUDE ) {
+                auto &attitude = current_tab[line].attitude;
+                switch( attitude ) {
+                    case Creature::A_HOSTILE:
+                        attitude = Creature::A_NEUTRAL;
+                        break;
+                    case Creature::A_NEUTRAL:
+                        attitude = Creature::A_FRIENDLY;
+                        break;
+                    case Creature::A_FRIENDLY:
+                        attitude = Creature::A_ANY;
+                        break;
+                    case Creature::A_ANY:
+                    default:
```

Comment:
The `default` case should be avoided when switching over an enumeration value. That allows the compiler to warn you when an enumeration value has not been handled (e.g. when somebody added a new value to the enumeration without updating the switch). The `default` case suppresses those warnings.